README.md
# MuonCutil

MuonCutil is a utility package for the MuonCollider software framework. It is intended to be a "meta-package" which packages together a set of independent utility packages living in separate sub-directories.

MuonCutil is distributed under the [GPLv3 License](http://www.gnu.org/licenses/gpl-3.0.en.html)

[![License](https://www.gnu.org/graphics/gplv3-127x51.png)](https://www.gnu.org/licenses/gpl-3.0.en.html)

## Usage

### confile
At the moment it cointains only some common configuration files for reconstruction


## License and Copyright
Copyright (C), iLCUtil Authors

iLCUtil is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License long with this program.  If not, see <http://www.gnu.org/licenses/>.


RELEASE_NOTES
#######################################################################################################################
v02-05-MC
#######################################################################################################################

External software

    Root 6.22.02-4 (Custom building)

Packages changes wrt. to v02-04-MC

    LCTuple → v01-13-MC
        Add information for tracker 
            Added track hits to the tracks branches
            Added the number of track hits

    lcgeo → v00-16-08-MC
        Yoke05_Barrel driver
            Removed slices DetElement construction
            Moved layer DetElement construction to external loop: fixes the DetElement hierarchy.
            This fixes the access to the DetElement given a CellID in our reconstruction. Related issue: iLCSoft/ILDConfig#88
        Add INSTALL_COMPACT_FILES option (default: OFF) to copy compact files to install area
        python3 compatibility
        For muoncollider:
            Fixed double thickness in the negative side of TrackerEndcapSupport
            Recovered the double thickness bug in TrackerEndcapSupport_o1_v01 for backward compatibility

    detector-simulation → v01-01-MC
        Added v2 of the frozen geometry with fixed endcap tracker support asymmetry

    DDMarlinPandora → v00-12-MC
        Added new processor (DDCaloDigi_BIB) for BIB subtraction in calorimeter cells, compatible with 1.5TeV BIB


#######################################################################################################################
v02-04-MC
#######################################################################################################################

External software

    Geant4 10.6.3

Packages changes wrt. to v02-03-MC

    Downgrade DD4Hep and DD4hepExamples to v01-13-01 as workaround for issue 759
    Add SIM.hepmc3.useHepMC3 = False to steering file for ddsim

#######################################################################################################################
v02-03-MC
#######################################################################################################################

External software

    ROOT_version = "6.22.02-3"

Packages changed wrt. to v02-02-MC

    MarlinTrkProcessors → v02-12-MC
        Double-layer filtering + Splitting by layers
            Introduces a new FilterDoubleLayerHits processor for filtering of RecHits based on their angular alignment within a double-layer.
            Updates the SplitCollectionByLayer processor to support custom layer combinations.
        Extra quality cuts in Refitter
            Minimal Chi2/NDF
            Minimal number of hits in arbitrary combinations of tracker subdetectors
    detector-simulation→ v01-00-MC
        Added a Python version of the MARS->SLCIO conversion script with configurable pdgIds
        Included antineutrons in Ekin cut. Fixed bug with cutting on Ekin before reading the MARS values
        Added reduced mod4 and mod5 geometries with Trackers only for faster simulation
        Added mod5 geometry with 3.57T magnetic field and rearranged VTX endcaps
        Added mod6 version of the geometry to be frozen for Snowmass → MuColl_v0: Copy of CLIC_o3_v14_mod4 with the MAP magnetic field
    ConformalTracking → v01-11-MC
        Added timing histograms for each tracking step
        Added support for excluding collections used for initial seed hit creation. Added per-step debug histograms
        Separated step debugging plots with an own configuration parameter 
    lcio →v02-15-04-MC
        fix nan's in jet 4-momenta (delphes2lcio)
        delphes: fix PDG in PID objects of neutral hadrons
        add optional extra PFO collections in delphes2lcio
        Python3 compatibility of Reader class 

Know problems

    The paths of the geometry in the MuonCutil/confile/* are wrongs.
        Use the correct ones: /opt/ilcsoft/v02-03-MC/detector-simulation/geometries/MuColl_v0/MuColl_v0.xml
    Default value for ddsim is to use HepMC3 Helper reader which is not supported by now. 
        Use the command line option for ddsim: --hepmc3.useHepMC3 false or put SIM.hepmc3.useHepMC3 = False in your steering file
    Using hepmc input file for ddsim produces a fatal exception likes this one: (see issue 759 in DD4Hep)

    G4DecayTable::SelectADecayChannel :: no possible DecayChannel mu+

    -------- EEEE ------- G4Exception-START -------- EEEE -------
    *** G4Exception : DECAY003
    issued by : G4Decay::DoIt
    Can not determine decay channel for mu+
    mass of dynamic particle: 0.00010566 (GEV)
    dacay table has 1 entries
    0: BR 1, IsOK? 0, --> e+ + nu_e + anti_nu_mu

    *** Fatal Exception *** core dump ***

        At the moment the only workaround is to downgrade the version of DD4Hep from v01-14 to v01-13-01

#######################################################################################################################
v02-02-MC
#######################################################################################################################

External software

    ROOT_version = "6.22.02"
    Boost_version = "1.69.0"
    Geant4 version = "10.6.1"
    clhep version = "2.4.1"

Official ILCSoft release v02-02 plus

    MarlinTrkProcessors → v02-11-MC
        Added time smearing to the DDPlanarDigiProcessor
        Changed histogram range suitable for picosecond-level time resolution
        Introduced a time window for digitized hits in the DDPlanarDigiProcessor
        Made TOF compensation persistent in the DDPlanarDigiProcessor to match the behaviour in DDCaloDigi
    lcgeo → v00-16-06-MC
        Added a modified version of the Generic Cal Endcap without cutout
        Added support of Z segmentation to ZPlanarTracker class
        Revoked ZPlanarTracker. Implemented Z segmentation in a separate plugin: ZSegmentedPlanarTracker
    ConformalTracking → v01-10-MC
        Add the possibility to do a regional tracking (define a theta range)

macros/make_trkPlots.C
#include "TFile.h"
#include "TTree.h"
#include "TEfficiency.h"
#include "TH1F.h"

/*
 =====================================================================================================================

 Simple macro to plot tracks efficiency. It produces:
    - Tracks efficiency vs pT
    - Tracks efficiency vs theta
    - Tracks efficiency vs phi
    - pT resoltution vs pT 
    - pT resoltution vs thetaT 

 If required, set debug = 1, plot also the base variables: pt, phi, theta, chi2 and nhit 
 If required, set saveonfile = 1, saved also the plots in png files.  

 Use: root -l make_trkPlots.C

   Copyright (C) 2020 Massimo Casarsa
 
 =====================================================================================================================
*/

// save plots on file?
const bool saveonfile = 1;
// plot also the base variables: pt, phi, theta, chi2 and nhit
const bool debug = 1;


TFile* infile;
TTree* T_eff;
TTree* T_trk;

TEfficiency* eff_pt;
TEfficiency* eff_phi;
TEfficiency* eff_theta;

TH1F* h_pt_reco;
TH1F* h_phi_reco;
TH1F* h_theta_reco;
TH1F* h_chi2_reco;
TH1F* h_nhit_reco;

const unsigned int Nbins_pt = 10;
TH1F* h_pt_res[Nbins_pt];

const unsigned int Nbins_theta = 20;
TH1F* h_theta_res[Nbins_theta];

TH1F h_pt_res_tot("h_pt_res_tot",";p_{T} [GeV];#sigma(#Deltap_{T}/p_{T}^{2}) [GeV^{-1}]",Nbins_pt,0.,10.);
TH1F h_theta_res_tot("h_theta_res_tot",";#theta [#circ];#sigma(#Deltap_{T}/p_{T}^{2}) [GeV^{-1}]",Nbins_theta,0.,90.);

TCanvas* c[10];
TCanvas* basevar[5];

// =====================================================================================================================

void make_trkPlots(TString filename="histograms.root"){

  // --- Book histograms:
  
  eff_pt = new TEfficiency("eff_pt",";p_{T} [GeV];#epsilon_{trk}",40,0.,10.);
  eff_phi = new TEfficiency("eff_phi",";#phi [#circ];#epsilon_{trk}",180,-180.,180.);
  eff_theta = new TEfficiency("eff_theta",";#theta [#circ];#epsilon_{trk}",60,0.,180.);

  h_pt_reco = new TH1F("h_pt_reco",";p_{T} [GeV]",100,0.,10.) ;
  h_phi_reco = new TH1F("h_phi_reco",";#phi [#circ]",120,-180.,180.) ;
  h_theta_reco = new TH1F("h_theta_reco",";#theta [#circ]",100,20.,150.) ;
  h_chi2_reco = new TH1F("h_chi2_reco",";#chi^{2}/ndf",100,0.,20.) ;
  h_nhit_reco = new TH1F("h_nhit_reco",";N_{hit}",20,0.,20.) ;

  for ( unsigned int ih=0; ih<h_pt_res_tot.GetNbinsX(); ++ih ) {
    TString hname = Form("h_pt_res_%d",ih);
    h_pt_res[ih] = new TH1F(hname,"",100,-0.01,0.01);
  }
  
  for ( unsigned int ih=0; ih<h_theta_res_tot.GetNbinsX(); ++ih ) {
    TString hname = Form("h_theta_res_%d",ih);
    h_theta_res[ih] = new TH1F(hname,"",100,-0.01,0.01) ;
  }
  

  // --- Open the root file and load the trees:

  infile = new TFile(filename);


  // Track reconstruction efficiency

  T_eff = (TTree*) infile->Get("MyClicEfficiencyCalculator/simplifiedEfficiencyTree");

  double pt_mc, phi_mc, theta_mc;
  int nhits;
  bool is_reco;
  T_eff->SetBranchAddress("m_pt", &pt_mc);
  T_eff->SetBranchAddress("m_theta", &theta_mc);
  T_eff->SetBranchAddress("m_phi", &phi_mc);
  T_eff->SetBranchAddress("m_nHits", &nhits);
  T_eff->SetBranchAddress("m_reconstructed", &is_reco);

  for(int ientry=0; ientry<T_eff->GetEntries(); ++ientry){

    T_eff->GetEntry(ientry);

    eff_phi->Fill(is_reco,phi_mc);
    eff_theta->Fill(is_reco,theta_mc);

    if (theta_mc<8. || theta_mc>172. ) continue;

    eff_pt->Fill(is_reco,pt_mc);

  }

    
  // Track parameters plots

  T_trk = (TTree*) infile->Get("MyClicEfficiencyCalculator/perfTree");
  
  std::vector<double> *pt_true=0;
  std::vector<double> *phi_true=0;
  std::vector<double> *theta_true=0;
  std::vector<double> *pt_reco=0;
  std::vector<double> *phi_reco=0;
  std::vector<double> *theta_reco=0;
  std::vector<double> *chi2_reco=0;
  std::vector<double> *nhit_reco=0;
  std::vector<double> *purity_reco=0;
  T_trk->SetBranchAddress("truePt", &pt_true);
  T_trk->SetBranchAddress("truePhi", &phi_true);
  T_trk->SetBranchAddress("trueTheta", &theta_true);
  T_trk->SetBranchAddress("recoPt", &pt_reco);
  T_trk->SetBranchAddress("recoPhi", &phi_reco);
  T_trk->SetBranchAddress("recoTheta", &theta_reco);
  T_trk->SetBranchAddress("recoChi2OverNDF", &chi2_reco);
  T_trk->SetBranchAddress("recoNhits", &nhit_reco);
  T_trk->SetBranchAddress("recoPurity", &purity_reco);

  for(unsigned int ientry=0; ientry<T_trk->GetEntries(); ++ientry){

    T_trk->GetEntry(ientry);

    for (unsigned int itrk=0; itrk<pt_reco->size(); ++itrk){

      h_pt_reco->Fill(pt_reco->at(itrk));
      h_phi_reco->Fill(phi_reco->at(itrk));
      h_theta_reco->Fill(theta_reco->at(itrk));
      h_chi2_reco->Fill(chi2_reco->at(itrk));
      h_nhit_reco->Fill(nhit_reco->at(itrk));

      // pt resolution
      
      double res_pt = (pt_reco->at(itrk)-pt_true->at(itrk))/(pt_true->at(itrk)*pt_true->at(itrk));

      int pt_bin = (int) ( pt_true->at(itrk)/h_pt_res_tot.GetBinWidth(1) );
      if ( pt_bin < 0 )
        pt_bin = 0;

      double theta = theta_true->at(itrk);
      if ( theta_true->at(itrk) > 90. )
	      theta = 180.-theta_true->at(itrk);
      int theta_bin = (int) ( theta/h_theta_res_tot.GetBinWidth(1) );

      h_pt_res[pt_bin]->Fill(res_pt);
      h_theta_res[theta_bin]->Fill(res_pt);
	
    }
      
  }

  // --- Fit the pt and theta bins and make the resolution plots:
  TF1* fitFunc = new TF1("fitFunc", "gaus");

  for ( unsigned int ih=0; ih<h_pt_res_tot.GetNbinsX(); ++ih ) {

    fitFunc->SetParameters(1.,h_pt_res[ih]->GetMean(), h_pt_res[ih]->GetRMS());
    
    h_pt_res[ih]->Fit("fitFunc","0E");

    h_pt_res_tot.SetBinContent(ih+1,fitFunc->GetParameter(2));
    h_pt_res_tot.SetBinError(ih+1,fitFunc->GetParError(2));
  
  }

  for ( unsigned int ih=0; ih<h_theta_res_tot.GetNbinsX(); ++ih ) {

    fitFunc->SetParameters(1.,h_theta_res[ih]->GetMean(), h_theta_res[ih]->GetRMS());
    
    h_theta_res[ih]->Fit("fitFunc","0E");

    h_theta_res_tot.SetBinContent(ih+1,fitFunc->GetParameter(2));
    h_theta_res_tot.SetBinError(ih+1,fitFunc->GetParError(2));
    
  }


  c[0] = new TCanvas("c_0","pT efficiency",800,600);
  eff_pt->Draw();
  c[0]->Update();
  // forces Y range to [0,1]
  eff_pt->GetPaintedGraph()->GetYaxis()->SetRangeUser(0., 1.05);
  if ( saveonfile )
    c[0]->Print("Eff_pt.png");

  c[1] = new TCanvas("c_1","phi efficiency",800,600);
  eff_phi->Draw();
  if ( saveonfile )
    c[1]->Print("Eff_phi.png");  
  
  c[2] = new TCanvas("c_2","theta efficiency",800,600);
  eff_theta->Draw();
  if ( saveonfile )
    c[2]->Print("Eff_theta.png");

  c[3] = new TCanvas("c_3","pT resolution vs pT",800,600);
  h_pt_res_tot.Draw();
  if ( saveonfile )
    c[3]->Print("Res_vs_pt.png");

  c[4] = new TCanvas("c_4","pT resolution vs theta",800,600);
  h_theta_res_tot.Draw();
  if ( saveonfile )
    c[4]->Print("Res_vs_theta.png");

  if ( debug ) {
    basevar[0] = new TCanvas("basevar_0","pT",800,600);
    h_pt_reco->Draw();
    if ( saveonfile )
      basevar[0]->Print("pt_reco.png");

    basevar[1] = new TCanvas("basevar_1","phi",800,600);
    h_phi_reco->Draw();
    if ( saveonfile )
      basevar[0]->Print("phi_reco.png");

    basevar[2] = new TCanvas("basevar_2","theta",800,600);
    h_theta_reco->Draw();
    if ( saveonfile )
      basevar[2]->Print("theta_reco.png");

    basevar[3] = new TCanvas("basevar_3","chi2",800,600);
    h_chi2_reco->Draw();
    if ( saveonfile )
      basevar[3]->Print("chi2_reco.png");

    basevar[4] = new TCanvas("basevar_4","nhit",800,600);
    h_nhit_reco->Draw();
    if ( saveonfile )
      basevar[4]->Print("nhit_reco.png");

  }

}


confile/simulation/sim_steer.py
import os

from DDSim.DD4hepSimulation import DD4hepSimulation
from g4units import mm, GeV, MeV, m, deg
SIM = DD4hepSimulation()

## The compact XML file
SIM.compactFile = "/opt/ilcsoft/muonc/detector-simulation/geometries/MuColl_v1/MuColl_v1.xml"
## Lorentz boost for the crossing angle, in radian!
SIM.crossingAngleBoost = 0.
SIM.enableDetailedShowerMode = True
SIM.enableG4GPS = False
SIM.enableG4Gun = False
SIM.enableGun = False
## InputFiles for simulation .stdhep, .slcio, .HEPEvt, .hepevt, .hepmc files are supported
SIM.inputFiles = ["muonGun_gen.slcio"]
## HepMC reader v3 is not supported
SIM.hepmc3.useHepMC3 = False
## Macro file to execute for runType 'run' or 'vis'
SIM.macroFile = []
## number of events to simulate, used in batch mode
SIM.numberOfEvents = 10
## Outputfile from the simulation,only lcio output is supported
# SIM.outputFile = "out/digi/sim_mod1_mumu2H2bb750.slcio"
SIM.outputFile = "muonGun_sim.slcio"
# SIM.outputFile = "out/sim_mod1_mcparticles.slcio"
## Verbosity use integers from 1(most) to 7(least) verbose
## or strings: VERBOSE, DEBUG, INFO, WARNING, ERROR, FATAL, ALWAYS
SIM.printLevel = 3
## The type of action to do in this invocation
## batch: just simulate some events, needs numberOfEvents, and input file or gun
## vis: enable visualisation, run the macroFile if it is set
## run: run the macroFile and exit
## shell: enable interactive session
SIM.runType = "batch"
## Skip first N events when reading a file
SIM.skipNEvents = 0
## Steering file to change default behaviour
SIM.steeringFile = None
## FourVector of translation for the Smearing of the Vertex position: x y z t
SIM.vertexOffset = [0.0, 0.0, 0.0, 0.0]
## FourVector of the Sigma for the Smearing of the Vertex position: x y z t
SIM.vertexSigma = [0.0, 0.0, 0.0, 0.0]


################################################################################
## Action holding sensitive detector actions
##   The default tracker and calorimeter actions can be set with
## 
##   >>> SIM = DD4hepSimulation()
##   >>> SIM.action.tracker = "Geant4TrackerAction"
##   >>> SIM.action.calo    = "Geant4CalorimeterAction"
## 
##   for specific subdetectors specific sensitive detectors can be set based on pattern matching
## 
##   >>> SIM = DD4hepSimulation()
##   >>> SIM.action.mapActions['tpc'] = "TPCSDAction"
## 
##   and additional parameters for the sensitive detectors can be set when the map is given a tuple
## 
##   >>> SIM = DD4hepSimulation()
##   >>> SIM.action.mapActions['ecal'] =( "CaloPreShowerSDAction", {"FirstLayerNumber": 1} )
## 
##    
################################################################################

##  set the default tracker action 
SIM.action.tracker = "Geant4TrackerWeightedAction" 

##  set the default calorimeter action 
SIM.action.calo = "Geant4ScintillatorCalorimeterAction"

##  create a map of patterns and actions to be applied to sensitive detectors
##         example: SIM.action.mapActions['tpc'] = "TPCSDAction" 
SIM.action.mapActions = {}


################################################################################
## Configuration for the magnetic field (stepper) 
################################################################################
SIM.field.delta_chord = 0.25*mm
SIM.field.delta_intersection = 0.001*mm
SIM.field.delta_one_step = 0.01*mm
SIM.field.eps_max = 0.001*mm
SIM.field.eps_min = 5e-05*mm
SIM.field.equation = "Mag_UsualEqRhs"
SIM.field.largest_step = 10.0*m
SIM.field.min_chord_step = 0.01*mm
SIM.field.stepper = "ClassicalRK4"


################################################################################
## Configuration for sensitive detector filters
## 
##   Set the default filter for tracker or caliromter
##   >>> SIM.filter.tracker = "edep1kev"
##   >>> SIM.filter.calo = ""
## 
##   Assign a filter to a sensitive detector via pattern matching
##   >>> SIM.filter.mapDetFilter['FTD'] = "edep1kev"
## 
##   Or more than one filter:
##   >>> SIM.filter.mapDetFilter['FTD'] = ["edep1kev", "geantino"]
## 
##   Don't use the default filter or anything else:
##   >>> SIM.filter.mapDetFilter['TPC'] = None ## or "" or []
## 
##   Create a custom filter. The dictionary is used to instantiate the filter later on
##   >>> SIM.filter.filters['edep3kev'] = dict(name="EnergyDepositMinimumCut/3keV", parameter={"Cut": 3.0*keV} )
## 
##    
################################################################################

##  default filter for calorimeter sensitive detectors; this is applied if no other filter is used for a calorimeter 
SIM.filter.calo = "edep0"

##  list of filter objects: map between name and parameter dictionary 
SIM.filter.filters = {'edep0': {'parameter': {'Cut': 0.0}, 'name': 'EnergyDepositMinimumCut/Cut0'}, 'geantino': {'parameter': {}, 'name': 'GeantinoRejectFilter/GeantinoRejector'}, 'edep1kev': {'parameter': {'Cut': 0.001}, 'name': 'EnergyDepositMinimumCut'}}

##  a map between patterns and filter objects, using patterns to attach filters to sensitive detector 
SIM.filter.mapDetFilter = {}

##  default filter for tracking sensitive detectors; this is applied if no other filter is used for a tracker
SIM.filter.tracker = "edep1kev"


################################################################################
## Configuration for the DDG4 ParticleGun 
################################################################################

##  direction of the particle gun, 3 vector 
# SIM.gun.direction = (0, 0, 1)

## choose the distribution of the random direction for theta
## 
##     Options for random distributions:
## 
##     'uniform' is the default distribution, flat in theta
##     'cos(theta)' is flat in cos(theta)
##     'eta', or 'pseudorapidity' is flat in pseudorapity
##     'ffbar' is distributed according to 1+cos^2(theta)
## 
##     Setting a distribution will set isotrop = True
##     
SIM.gun.distribution = "uniform"
SIM.gun.energy = 10.0*GeV

##  isotropic distribution for the particle gun
## 
##     use the options phiMin, phiMax, thetaMin, and thetaMax to limit the range of randomly distributed directions
##     if one of these options is not None the random distribution will be set to True and cannot be turned off!
##     
SIM.gun.isotrop = True
SIM.gun.multiplicity = 10
SIM.gun.particle = "mu-"
SIM.gun.phiMax = None

## Minimal azimuthal angle for random distribution
SIM.gun.phiMin = None

##  position of the particle gun, 3 vector 
SIM.gun.position = (0.0, 0.0, 0.0)
SIM.gun.thetaMax = 172*deg
SIM.gun.thetaMin = 8*deg


################################################################################
## Configuration for the output levels of DDG4 components 
################################################################################

## Output level for input sources
SIM.output.inputStage = 3

## Output level for Geant4 kernel
SIM.output.kernel = 3

## Output level for ParticleHandler
SIM.output.part = 3

## Output level for Random Number Generator setup
SIM.output.random = 6


################################################################################
## Configuration for the Particle Handler/ MCTruth treatment 
################################################################################

##  Keep all created particles 
SIM.part.keepAllParticles = False

## Minimal distance between particle vertex and endpoint of parent after
##     which the vertexIsNotEndpointOfParent flag is set
##     
SIM.part.minDistToParentVertex = 2.2e-14

## MinimalKineticEnergy to store particles created in the tracking region
SIM.part.minimalKineticEnergy = 1.0*MeV

##  Printout at End of Tracking 
SIM.part.printEndTracking = True

##  Printout at Start of Tracking 
SIM.part.printStartTracking = True

## List of processes to save, on command line give as whitespace separated string in quotation marks
SIM.part.saveProcesses = ['Decay']


################################################################################
## Configuration for the PhysicsList 
################################################################################
SIM.physics.decays = False
SIM.physics.list = "QGSP_BERT"

##  location of particle.tbl file containing extra particles and their lifetime information
##     
SIM.physics.pdgfile = os.path.join( os.environ.get("DD4HEP"), "DDG4/examples/particle.tbl" )

##  The global geant4 rangecut for secondary production
## 
##     Default is 0.7 mm as is the case in geant4 10
## 
##     To disable this plugin and be absolutely sure to use the Geant4 default range cut use "None"
## 
##     Set printlevel to DEBUG to see a printout of all range cuts,
##     but this only works if range cut is not "None"
##     
SIM.physics.rangecut = 0.7*mm

SIM.physics.rejectPDGs = {1,2,3,4,5,6,21,23,24,25}

################################################################################
## Properties for the random number generator 
################################################################################

## If True, calculate random seed for each event based on eventID and runID
## allows reproducibility even when SkippingEvents
SIM.random.enableEventSeed = True
SIM.random.file = None
SIM.random.luxury = 1
SIM.random.replace_gRandom = True
SIM.random.seed = None
SIM.random.type = None



releases/development/release-versions.py
###########################################
#
# iLCSoft versions for installing a current HEAD
# version of the ilcsoft packages.
# The external base tools need to be installed
#
# DESY ilcsoft team
###########################################
import datetime
import platform

# --------- ilcsoft release version ------------------------------------------
ilcsoft_release='muonc'
# ----------------------------------------------------------------------------

#-----------------------
# we now always build with c++11 ?
# use_cpp11 = True
# if nightlies:
#    use_cpp11 = nb_use_cpp11
#    print "******************* use_cpp11", use_cpp11

# which cxx standard to use
cxx_standard = 14

afsPath = None
try:
    afsPath = ilcsoft_afs_path[ arch ]
except KeyError:
    pass

#===============================================================================
# use a compiler that knows c++11, run
#
'''
#===== gcc 4.9:
# . ./scripts/use_gcc49_afs.sh
 source /afs/cern.ch/sw/lcg/contrib/gcc/4.9.3/x86_64-slc6/setup.sh
 export PATH=/afs/cern.ch/sw/lcg/releases/LCG_87/Python/2.7.10/x86_64-slc6-gcc49-opt/bin:$PATH
 export LD_LIBRARY_PATH=/afs/cern.ch/sw/lcg/releases/LCG_87/Python/2.7.10/x86_64-slc6-gcc49-opt/lib:$LD_LIBRARY_PATH


# or on cvmfs:
# . ./scripts/use_gcc49_cvmfs.sh
 source /cvmfs/sft.cern.ch/lcg/contrib/gcc/4.9.3/x86_64-slc6/setup.sh
 export PATH=/cvmfs/sft.cern.ch/lcg/releases/LCG_87/Python/2.7.10/x86_64-slc6-gcc49-opt/bin:$PATH
 export LD_LIBRARY_PATH=/cvmfs/sft.cern.ch/lcg/releases/LCG_87/Python/2.7.10/x86_64-slc6-gcc49-opt/lib:$LD_LIBRARY_PATH
'''
# before starting the installation
#================================================================================



# --------- install dir ------------------------------------------------------
# ===========================================================
# Modify this path to where you want to install the software
# ===========================================================
#
if installPrefix is None:
    ilcsoft_install_prefix = ilcsoft_afs_path[ arch ]
else:
    ilcsoft_install_prefix = installPrefix
#ilcsoft_install_prefix = "/cvmfs/ilc.desy.de/sw/x86_64_gcc48_sl6/"
#ilcsoft_install_prefix = "/scratch/ilcsoft/"

# ----------------------------------------------------------------------------
#--- the ilcsoft_release is now automatically appended in release-ilcsoft.cfg
#     but not in release-base.cfg !!

#append_version_to_install_prefix = False
#if(append_version_to_install_prefix):
#    ilcsoft_install_dir = os.path.join(ilcsoft_install_prefix , ilcsoft_release )

# ----------------------------------------------------------------------------


# --------- ilcsoft home -----------------------------------------------------
# ===========================================================
# Modify this path to where you want ilcinstall to look
# for pre-installed (base) packages
# typically this would be left to ilcsoft_install_prefix
# or set to an /afs or /cvmfs base installation that you
# want to use
# ===========================================================

ilcPath = ilcsoft_install_prefix
# ----------------------------------------------------------------------------


#--------------------------------------------------------------------------
#ilcPatchPath = "/afs/desy.de/project/ilcsoft/sw/x86_64_gcc41_sl5/v01-15"



# ======================= PACKAGES WITH NO INSTALL SUPPORT ===================

# these packages need to be pre-installed for your system
# please adjust the path variables accordingly

# detect the default software installation path
# when using package manager like apt-get, yum or brew
platfDefault = None

if platform.system().lower().find('linux') >= 0:
   platfDefault = '/usr'
elif platform.system().lower().find('darwin') >= 0:
   platfDefault = '/usr/local'

# ----- mysql --------------------------------------------------------
MySQL_version = "5.0.45"
MySQL_path = platfDefault

# overwrite with a patch set in the environment
my_mysql_path = os.getenv("MYSQL_DIR", default=None)
if( my_mysql_path !=  None ):
    MySQL_path = my_mysql_path


#MySQL_path = ilcPath + "/mysql/" + MySQL_version

##########################################################################################
#
#  end of user configuration section
#  only make changes below if you know what you are doing ...
#
##########################################################################################


# ======================= PACKAGE VERSIONS ===================================

Geant4_version =  "10.04.p03"

CLHEP_version =  "2.3.4.3"

ROOT_version = "6.22.02"

GSL_version = "2.6"

Qt5_version = "v5.13.1"

CMake_version = "3.15.5"

FastJet_version = "3.2.1"
FastJetcontrib_version = "1.025"

CED_version = "v01-09-03"

SIO_version = "v00-00-02"

Boost_version = "1.69.0"

Eigen_version = "3.3.7"

# -------------------------------------------

LCIO_version = "v02-15-04-MC"

GEAR_version = "v01-09"

CondDBMySQL_version = "CondDBMySQL_ILC-0-9-7"

ILCUTIL_version = "v01-06-01"

MarlinFastJet_version = "v00-05-02"


# -------------------------------------------

KalTest_version = "v02-05"

KalDet_version = "v01-14-01"

aidaTT_version = "v00-10"

DDKalTest_version = "v01-06"

MarlinTrk_version = "v02-08"

MarlinTrkProcessors_version = "v02-12-MC"

Clupatra_version = "v01-03"

KiTrack_version = "v01-10"

KiTrackMarlin_version = "v01-13"

FCalClusterer_version = "v01-00-01"

ForwardTracking_version = "v01-14-MC"

ConformalTracking_version = "v01-11-MC"

LICH_version = "v00-01"

# -------------------------------------------

GBL_version = "V02-02-00" #"V02-00-00"

LCCD_version = "v01-05"

RAIDA_version = "v01-09"

MarlinUtil_version = "v01-15-01"

Marlin_version = "v01-17-01"

MarlinDD4hep_version = "v00-06"

MarlinReco_version = "v01-28"

ILDPerformance_version = "v01-08"

#ILDConfig_version = "HEAD"


LCFIVertex_version = "v00-08"
LCFIPlus_version = "v00-09"


MarlinKinfit_version = "v00-06"
MarlinKinfitProcessors_version = "v00-04-02"

PandoraPFANew_version   = "v03-13-02"
DDMarlinPandora_version = "v00-12-MC"
PandoraAnalysis_version = "v02-00-01"

CEDViewer_version = "v01-17-01"

Overlay_version = "v00-22-02-MC"

PathFinder_version = "v00-06-01"

MarlinTPC_version = "v01-07"

LCTuple_version = "v01-13-MC"

BBQ_version = "v00-01-04"

Druid_version = "2.2"

Garlic_version = "v03-01"

DD4hep_version = "v01-13-01"
DD4hepExamples_version = "v01-13-01"

lcgeo_version = "v00-16-08-MC"

Physsim_version = "v00-04-01"


# xerces-c (needed by geant4 for building gdml support - required by mokka)
XercesC_version = "v3.2.2"
XERCESC_ROOT_DIR = ilcPath + "/xercesc/" + XercesC_version
